File: sandbox/task_iir_lowpass.c
#include <stdio.h>
#include <math.h>

#define PI 3.14159265358979323846

float iir_lowpass_filter(float input, float cutoff_freq, float sample_rate, float *prev_output) {
    float alpha = 2 * PI * cutoff_freq / (2 * PI * cutoff_freq + sample_rate);
    float output = alpha * input + (1 - alpha) * (*prev_output);
    *prev_output = output;
    return output;
}

int main() {
    float sample_rate = 1000.0f;
    int num_samples = 10;
    
    float fc100_prev = 0.0f;
    float fc150_prev = 0.0f;
    
    printf("Sample 0: raw=0.000, fc100=0.000, fc150=0.000\n");
    
    for (int i = 1; i < num_samples; i++) {
        float t = i / sample_rate;
        float raw_signal = sin(2 * PI * 50 * t) + sin(2 * PI * 200 * t);
        
        float fc100_out = iir_lowpass_filter(raw_signal, 100.0f, sample_rate, &fc100_prev);
        float fc150_out = iir_lowpass_filter(raw_signal, 150.0f, sample_rate, &fc150_prev);
        
        printf("Sample %d: raw=%.3f, fc100=%.3f, fc150=%.3f\n", 
               i, raw_signal, fc100_out, fc150_out);
    }
    
    return 0;
}

File: sandbox/task_iir_lowpass.py
import math

PI = 3.1415926535

def iir_lowpass(input_val, prev_output, alpha):
    output = alpha * input_val + (1.0 - alpha) * prev_output
    return output, output

def main():
    fs = 100.0
    N = 100
    t = [i / fs for i in range(N)]
    signal = [math.sin(2.0 * PI * 1.0 * t[i]) + 0.3 * math.sin(2.0 * PI * 10.0 * t[i]) for i in range(N)]
    
    prev_output1 = 0.0
    prev_output2 = 0.0
    filtered1 = []
    filtered2 = []
    
    for i in range(N):
        output1, prev_output1 = iir_lowpass(signal[i], prev_output1, 0.1)
        output2, prev_output2 = iir_lowpass(signal[i], prev_output2, 0.5)
        filtered1.append(output1)
        filtered2.append(output2)
    
    print('t=0.00, raw=0.000, alpha0.1=0.000, alpha0.5=0.000')
    for i in range(10):
        print(f't={t[i]:.2f}, raw={signal[i]:.3f}, alpha0.1={filtered1[i]:.3f}, alpha0.5={filtered2[i]:.3f}')

if __name__ == '__main__':
    main()